### ISSUE TITLE
[BUG] Spellqueue doesnt work if heroic strike or Cleave is queued.

### ISSUE THREAD

**Issue by Infurnalz** [JamminL/wotlk-classic-bugs#572]
#### 1.1. Description
The way that spellqueue currrently works for warriors, is that as long as HS/cleave is not queued you can queue up your next ability correctly. However once HS is queued you are unable to do so. Resulting in no spellqueue window, which increases delay on abilities and therefore lowering dps. 
This forces the warrior to spam abilities to counter some of the delay, but it will still result in some delay depending on human factors and your own latency. 

Display of it not working.

https://streamable.com/38p7j6

https://streamable.com/g8kbue

There has also been other mentions of this.

https://us.forums.blizzard.com/en/wow/t/this-bug-is-one-of-the-reasons-why-fury-warriors-are-so-bad-at-the-moment/1408439

https://us.forums.blizzard.com/en/wow/t/warrior-spell-queue-not-working/1421625

  #### 1.2. How to Reproduce
  To reproduce it does requires you to attempt several times, as we are talking about a small window of where you can queue your ability and therefore it is not always transparent when you are using the spellqueue system.
     
       1. Ensure that spellqueuewindow is set to the default value of 400 
       2. Gain a full rage bar.
       3. Attack a dummy and use Heroic Strike (or Cleave) right after your main hand has struck the target~.
       4. Use an ability that's on the GCD, for example Bloodthirst.
       5. Once the GCD cooldown timer is below the spellqueuwindow (400ms in this example) press another ability on the GCD, whirlwind.
       6. Once the GCD cooldown is over, the next ability should be cast, but it doesnt.
       7. Repeat this process without using Heroic Strike (or Cleave) and notice how Whirlwind does fire.
      
  #### 1.3  Source Material
  The red bar is the GCD of the ability and I am casting BT and WW.
  https://streamable.com/1warwp
  
  https://streamable.com/shr169
  
  https://streamable.com/90hpo4
  
  Here we see it working without HS
  
  https://streamable.com/g8kbue
  
https://www.youtube.com/watch?v=l--e74c9v48&t=21s

Video of Cleave showing the same bug
https://streamable.com/38p7j6

#### 2.1. Description
The spell queue functions as a system for lag mitigation, allowing users to queue an ability while waiting for GCD cooldown to come off. This removes the local delay and system delay one would typically get. The spellqueue window can be set from 0-400, and 400 is the default that most players are using. The spellqueue window timing can be found using:
```
/dump GetCVar("SpellQueueWindow")
```
It is a system that benefits everyone, as no matter how good your system is you will always have different degree's of delay. 

The expected behaviour is therefore that when you use a GCD ability, while having HS queued, you would expect that you could queue up your next ability and it would cast. However if HS/cleave is queued while this happens, it does not work, forcing you to spam buttons and experience delay in your casts. Resulting in a decent dps loss. 

Older videos seem to display that it was not a bug back then, but as video quality is bad, and the systems generally had more latency it can be hard to determine.

  #### 2.2 Source Material
  http://thedefiantguild.com/forums/discussion/15431.html
  
  https://streamable.com/g8kbue
  
  https://www.youtube.com/watch?v=OQwT1gYSDVk&t=50s
  
  https://youtu.be/jQxQ7TJQJQg

**Comment by NuptupTDOW** [comment]
Can someone check if the same thing happens with rune strike for death knights?

**Comment by Infurnalz** [comment]
It seems that abilities that are queued to next melee attack is affected, meaning if you were to queue Heroic Strike the spellqueue would'nt work despite heroic strike being queued outside of the default set spellqueue timing : 400ms.

Im not entirely sure how much this affects different classes, but for warriors as we have heroic strike or cleave queued almost the entire time, the spellqueue system is basically fully removed. 

Not sure what research is needed in terms of determining its validity, but it is pretty evident this is not intentional game design.
There could be discussion about whether the command /cqs (cancelspellqueue) could be used to avoid this, though im not sure if it even works and if it did how it would affect abilities such as heroic strike when they are queued.

**Comment by CharWS** [comment]
Still don't see any information about how it used to work originally to confirm whether the interaction is authentic or not

**Comment by knasen88** [comment]
What do you mean? Queued spells like heroic strike was removed before spell queueing entered the game, to not have it work for warriors is just silly

**Comment by Infurnalz** [comment]
> Still don't see any information about how it used to work originally to confirm whether the interaction is authentic or not

The current spell queue system wasnt a part of wrath. You have quite a few mentions of it when it was introduced.
http://tobolds.blogspot.com/2010/10/wow-patch-punishes-button-mashing.html
https://gamefaqs.gamespot.com/boards/534914-world-of-warcraft/56749820
https://gamefaqs.gamespot.com/boards/534914-world-of-warcraft/56737426
With the 4.3 patch the lag tolerance slider was introduced.
And at that point heroic strike was changed "is no longer a On-Next-Swing ability, but instead is a normally activated attack."
And that isnt the current spellqueue system either as it was changed during legion.

Even so, while Wotlk classic is trying to imitate the original wrath release, it is running on a different system. The iteration of the current spellqueue system is not like it was back when it was introduced in cata. So I dont see much reason to discuss whether its authentic wrath behavior.

It wasnt a system that was implemented back then, and it is completely illogical to queue an ability outside of the spellqueuewindow, and then it disables any other interaction with the spellqueuewindow till the next melee hit is processed. 

Just as extra information: during vanilla, tbc and wrath(in some interactions) people used to do stopcasting (the equivalent of the spellqueue system back then) to gain a competitive edge. This was later removed and changed into the "lag tolerance" and later spellqueue system. The stopcasting mechanic was also used for a wide range of interactions no longer possible such as /slam /stopcasting allowing you to essentially just spam slam regardless of procs, and it would only interact when bloodsurge is up. 
Most of the older macros made during this era no longer works due to blizzard changing the systems. 
https://wowwiki-archive.fandom.com/wiki/Useful_macros_for_warriors

Also during wrath and later, it was not uncommon for people to "mod" their registry to try to increase frequency in packets to lower their delay also known as Leatrix Latency Fix. 
https://www.wowhead.com/forums/topic/wanna-speed-up-your-latency-57226
https://learn.microsoft.com/da-DK/troubleshoot/windows-server/networking/registry-entry-control-tcp-acknowledgment-behavior (probably in danish)

To end off I want to look at the introduction of lag tolerance as a whole.
"Client spell cast requests are now sent to the server even if your player is already casting another spell" was introduced in 2.3.0 (zul aman) which allows you to send spell cast requests at any point. This was to minimize the delay between pc -> server -> pc illustrated in this picture https://wowhats.files.wordpress.com/2010/03/clientservercasting.png 
Blizzard then allowed spell casts requests to be sent at anytime (though you couldnt go below the 1sec gcd timer) effectively minimizing most the delay you had. 

I have browsed plenty of videos looking for an example to display that it either was or wasnt a thing back then, but the problem is that with lower frames, lower reaction times, higher latency in equipment and high latency in internet speed both in pc to server but also server to pc. The results of that would be quite muddy. And even so, at that point they had different commands availible that probably could remove it. 

Long post
TLDR: It is not the same spell queue system that was back then, so it doesnt really apply.

**Comment by Infurnalz** [comment]
After checking my source material again the video linked in the original post shows points where it seems to work as it should compared to how it is now.
Source: https://www.youtube.com/watch?v=jQxQ7TJQJQg 0:36
Though this is again difficult to see since we are talking low framerate video and high delay already.

**Comment by JamminL** [comment]
To shed some light, you are correct @Infurnalz. Lag tolerence, as we know it today, was introduced in cata. Wotlk had another version of it but nothing that would allow IN-GCD queueing.

What this means for this bug report is that we have a mechanic that was backported to classic that works with every spell except when queued spells are involved. This, in my opinion, is bug-worthy.

It would be interesting to see how it behaves in 10.0, if it's a problem with wotlk-specific queued spells or retail lag tolerence also doesn't handle this type of queued spells.

**Comment by JamminL** [comment]
After checking I don't think there's any spell that works like heroic strike or cleave on retail, so that's probably part of the reason why this doesn't work.

**Comment by JamminL** [comment]
You could in theory check warrior cata videos because if I recall correctly heroic strike was still a thing back then, and so was Lag Tolerence / spell queue window. But it'd be a tough thing to find

**Comment by Infurnalz** [comment]
Problem with checking with cata videos is that the way heroic strike is interpreted was changed "Heroic strike is no longer a On-Next-Swing ability, but instead is a normally activated attack." I also think that this in part is where the issue lies, since it could be considered an ability. I would assume that all on next swing melee attacks were changed at that time. Though that it mostly based on my memories and "bad" blackbox testing.


However I did find the bluepost for the change and some classes are mentioned.
"Heroic Strike will no longer be an "on next swing" attack, as we are removing this mechanic in Cataclysm. To keep the niche of Heroic Strike as a Rage dump, it will become an instant attack....Other abilities, such as Cleave, Execute, and Maul (for druids) will work similarly. The goal is to provide players with an option where if you can't afford the Rage, you don't push the button, but if you have excess Rage, you can push it a lot."
https://www.bluetracker.gg/wow/topic/us-en/24038612592-cataclysm-class-preview-warrior/

**Comment by sjsctt1** [comment]
it has been found that /stopcasting can indeed be used to enable spellqueue while casting heroic strike. 
[https://cdn.discordapp.com/attachments/1013562289678467134/1095851362664861826/image.png](url)
It also still works if you spam your key
[https://cdn.discordapp.com/attachments/1013562289678467134/1095851443094814742/2023-04-13_06-09-06.mp4](url)
It can be assumed that once a spell is in the spellqueue, heroic strike would not remove the existing spell in the spellqueue

**Comment by Miyur1** [comment]
https://www.wowhead.com/blue-tracker/topic/eu/wrath-classic-patch-3-4-2-update-notes-may-11-446339
- Fixed an issue that caused On Next Hit melee attacks to disable the spell queue window.
